R/sampler-hmc.R

```diff
@@ -0,0 +1,220 @@
+    v_next <- v_next + epsilon * compute_gradient(pars_next) / 2
+
+    ## Some treatments would negate momentum at end of trajectory to
+    ## make the proposal symmetric by setting v_next <- -v_next but
+    ## this is not actually needed in practice.
+
+    ## Potential energy at the end of the trajectory
+    density_next <- model$density(pars_next)
+
+    ## Kinetic energies at the start (energy) and end (energy_next) of
+    ## the trajectories
+    energy <- sum(v^2) / 2
+    energy_next <- sum(v_next^2) / 2
+
+    ## Accept or reject the state at end of trajectory, returning
+    ## either the position at the end of the trajectory or the initial
+    ## position
+    u <- rng$random_real(1)
+    accept <- u < exp(density_next - state$density + energy - energy_next)
+    if (accept) {
+      state$density <- density_next
+      state$pars <- pars_next
+    }
+
+    if (debug) {
+      internal$history <-
+        c(internal$history, list(list(pars = history, accept = accept)))
+    }
+
+    state
+  }
+
+  finalise <- function(state, model, rng) {
+    if (debug) {
+      pars <- lapply(internal$history, "[[", "pars")
+      pars <- array(unlist(pars),
+                    c(dim(pars[[1]]), length(pars)),
+                    list(NULL, model$parameters, NULL))
+      accept <- vlapply(internal$history, "[[", "accept")
+      list(debug = list(pars = pars, accept = accept))
+    } else {
+      NULL
+    }
+  }
+
+  mcstate_sampler("Hamiltonian Monte Carlo",
+                  initialise,
+                  step,
+                  finalise)
+}
+
+
+hmc_transform <- function(domain) {
+  lower <- domain[, 1]
+  upper <- domain[, 2]
+
+  is_bounded <- is.finite(lower) & is.finite(upper)
+  is_semi_infinite <- is.finite(lower) & upper == Inf
+  is_infinite <- lower == -Inf & upper == Inf
+  unknown <- !(is_bounded | is_semi_infinite | is_infinite)
+  if (any(unknown)) {
+    ## This is not very likely to be thrown
+    cli::cli_abort("Unhandled domain type for parameter {which(unknown)}")
+  }
+
+  ## Save finite bounds for later:
+  a <- lower[is_bounded]
+  b <- upper[is_bounded]
+
+  list(
+    ## Transform from R^n into the model space
+    rn2model = function(theta) {
+      theta[is_semi_infinite] <- exp(theta[is_semi_infinite])
+      theta[is_bounded] <- ilogit_bounded(theta[is_bounded], a, b)
+      theta
+    },
+
+    ## Transform from model space to R^n
+    model2rn = function(x) {
+      x[is_semi_infinite] <- log(x[is_semi_infinite])
+      x[is_bounded] <- logit_bounded(x[is_bounded], a, b)
+      x
+    },
+
+    ## Derivative of rn2model, as a multiplier to f'(x); we pass in
+    ## model parameters here, not R^n space.
+    deriv = function(x) {
+      ## d/dtheta theta -> 1
+      x[is_infinite] <- 1
+      ## Here we want
+      ## d/dtheta f(exp(theta)) -> exp(theta) f'(exp(theta))
+      ##                        -> x f'(x)
+      x[is_semi_infinite] <- x[is_semi_infinite]
```

Comment:
Not sure I see what this line means

Reply:
that line can be removed, but is in there to remind me that's what we're doing really! Probably this would be clearer if we assigned into something other than `x`; I'll tweak that